docs: add a next-phase CRD design document

phlogistonjohn opened the pull request:
I've been working on this in the background and I think it's good enough to share now. I do not expect to merge this as-is but to mainly use it to kick off additional discussion. That said, I think it would be good to merge it once we agree on the broad strokes and then continue to modify it there.

Please, do ask questions and make me justify some of these ideas! :-)

Part of: #2

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,215 @@
+instance.
+
+SmbShare CRs will refer to one or more SmbSecurityDomain resources. The
+SmbSecurityDomain refernce will define the main security properties of the
```
`refernce` ==> `reference`

phlogistonjohn replied:
should be fixed.

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,215 @@
+
+Spec Options:
+* `shareName` - string - Optional string giving the share a full SMB compliant name. If unset, the name will be derived by the operator.
+* `storage` - mapping - A new subsection to configure the storage backing the share.  * `pvc` - mapping - A new subsection used to configure PVC backing storage.
```
newline missing

phlogistonjohn replied:
fixed.

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,215 @@
+    * `spec` - mapping, embedded pvc spec - An embedded PVC spec that will be used to dynamically create a backing PVC for the share; sharing the lifecycle of the PVC with the share
+  * TBD - Any other more custom storage backends if needed
+* `securityDomain`  - string - The name of the SmbSecurityDomain CR associated with this share
+* `grouping` - string - Optional string that if undefined will be generated by the operator. The grouping is used by the operator as a hint to combine shares under one smbd service, depending on other settings
```
What do you imagine the value of `grouping`to look like? A list of other smbshare names?

obnoxxx replied:
OK, I see it in the examples below. Just a group name that will have all the shares with the same grouping name combined under on smbd instance.

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,215 @@
+* `grouping` - string - Optional string that if undefined will be generated by the operator. The grouping is used by the operator as a hint to combine shares under one smbd service, depending on other settings
+* `clustering` - mapping - A new subsection pertaining to clustering smbd instances
+  * `mode` - enumerated string "none" or "ctdb" - This option tells the operator if this share should be clustered or not
+  * `minSize` - int - Minimum number of smbd instances when clustering
```
Why minimum and not an exact size?

phlogistonjohn replied:
I wasn't sure if we'd want to have a minSize and a maxSize in the future. But perhaps that is premature. Would you prefer if I changed it to just "size" and we can deal with the future in the future?

obnoxxx replied:
Note from earlier discussion: minSize was chosen to allow for more flexible grouping...

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,215 @@
+  * TBD - Any other more custom storage backends if needed
+* `securityDomain`  - string - The name of the SmbSecurityDomain CR associated with this share
+* `grouping` - string - Optional string that if undefined will be generated by the operator. The grouping is used by the operator as a hint to combine shares under one smbd service, depending on other settings
+* `clustering` - mapping - A new subsection pertaining to clustering smbd instances
```
We might want to structure this a little differently, not yet sure how: As mentioned I was identifying 3 aspects of samba's dbwrap backend currently provided by ctdb
1. indication of cluster intactness in the backend (recovery lock file)
2. volatile DBs (and messaging)
3. persistent DBs.

of these, #1 might be useful without clustering, #3 is definitely useful for making config and settings (join...) persistent. But #2 would really for moving to all-active clustering. This would require development in samba though. 

So to summarize, not sure if putting this under "clustering" is entirely sufficient or appropriate.

phlogistonjohn replied:
Sure. At the time I was only thinking of non-clustered or clustered. I do think that some sort of subsection would be needed to what kind of deployment (in the general sense, not just the k8s resource) we want and how it is expected to scale. Perhaps if we replace "clustering" with "scaling" as a more generic term? And then the type of scaling might be single instance or clustered or some hybrid...

phlogistonjohn replied:
Also, I forgot to note. None of this is discussing what the defaults would be unless explicitly stated. I suspect that in some cases we might prefer to have the more "complex" mode the default but allow some use cases to "opt-out". For example if a clustered mode was the default, a user might want to scale down to one instance for a read only share of limited importance to save on resources, etc.

obnoxxx reviewed (commented):
In general, I like the idea of structuring it this way very much. A couple of comments and thoughts inside.

phlogistonjohn commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,215 @@
+* `realm` - string - The dns domain name of an Active Directory domain
+* `domains` - list-of-maps - A new subsection pertaining to detailed "active-directory" configuration (optional)
+  * `backend` - enumerated string - the name of a (supported) ID mapping backend
+  * `domainName` - string - the name of the domain being configured
```
The `realm` translates to the smb.conf parameter of the same name. The `domainName` keys loosely map to the domain key of the `idmap` smb.conf settings. These are optional but allow one to say configure a non-default idmap setting like "ad" for either the primary `realm` or other trusted domains.

obnoxxx replied:
Got it.

phlogistonjohn commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,215 @@
+* `localName` - string - Optional string used to indicate how this instance will
+  be referred to in an Active Directory
+* `realm` - string - The dns domain name of an Active Directory domain
+* `domains` - list-of-maps - A new subsection pertaining to detailed "active-directory" configuration (optional)
```
It is supposed to contain as many settings for domains as you have (primary or trusted) that you want to configure.

Example:
```
domains:
  - domainName: myrealm.mycorp.example.com
    backend: ad
  - domainName: atrust.mycorp.example.com
    backend: ad
```

Do you you think it would be clearer as a map-of-maps like this example:
```
domains:
    myrealm.mycorp.example.com:
        backend: ad
    atrust.mycorp.example.com:
        backend: ad
```?

If we need to the section could include other parameters such as idmap ranges, etc. But I left it out of the initial draft becase (a) I was hoping to keep this siimple and (b) hoping the operator could do the heavy lifting of determining valid ranges after being simply told to use the ad (and check for rfc2307 in the ad).

obnoxxx replied:
I think a tag for the name is fine, but `domainName` under `domains` is kind of redundant. we could just have `name` under `domains`?

phlogistonjohn replied:
Sure. I thought a more descriptive key would be better but it is somewhat redundant.

phlogistonjohn replied:
changed.

phlogistonjohn commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,215 @@
+The operator may, or may not, combine one or more share into a single smbd
+instance.
+
+SmbShare CRs will refer to one or more SmbSecurityDomain resources. The
```
It would not. This is poorly worded. I will improve it.

phlogistonjohn replied:
updated.

phlogistonjohn commented:
Pushed an updated version that fixes some of the items that didn't warrant additional discussion. I'm looking forward to hearing your opinions on the remaining items!

obnoxxx reviewed (commented):
After your comments and thinking about this a bit more, I think we should rename `SmbSecurityDomain`: This fundamentally corresponds to the global configuration of a samba server (`smb.conf` file). The `SmbShare` type corresponds to a share section in an `smb.conf` file. 

So in this regard, what you named `SmbSecurityDomain` might better be named `SambaInstance` or `SambaServer`. The instance is ultimately identified by its internal ID (the machine SID) once the server/db is initialised. The fqdn (essentially `localName` + `realm`) can also identify the instance.

There can only be one main smbd for a given SID / fqdn. So the `SmbShare` CRs need to end up as share config sections for their respective smbd instances.

Each `SmbShare` needs to be associated to one and only one `SambaServer` (or `SambaSecurityDomain`). Also very important that (1) each PVC is only attached to only one `SambaServer` (indirectly). If there is a possibility for a subdirectory of a PVC to be turned into another PVC, then it is important that these two PVCs are attached to the *same* `SambaServer`.

phlogistonjohn commented:
> After your comments and thinking about this a bit more, I think we should rename `SmbSecurityDomain`: This fundamentally corresponds to the global configuration of a samba server (`smb.conf` file). The `SmbShare` type corresponds to a share section in an `smb.conf` file.
> 

I agree that the security & domain settings reside in the global section of the smb.conf. I also agree that SmbShare aligns with the share section of smb.conf.


> So in this regard, what you named `SmbSecurityDomain` might better be named `SambaInstance` or `SambaServer`. The instance is ultimately identified by its internal ID (the machine SID) once the server/db is initialised. The fqdn (essentially `localName` + `realm`) can also identify the instance.
> 

I was trying to explicitly avoid tying the operator's config to either server instances or smb.conf. The names of the many config items reflect the parameters in smb.conf my intention is not to pass these options to smb.conf in a  1:1 manner. Rather the names are used for familiarity.  

I prefer SmbSecurityDomain because:
* It only covers the security and domain settings, while real smb.conf has many many more global options
* It emphasizes the important user provided nature of security settings
* It manages some important security and domain settings that typically lie outside of smb.conf - mainly the reference to a secret holding user information for non-domain-joined mode


> There can only be one main smbd for a given SID / fqdn. So the `SmbShare` CRs need to end up as share config sections for their respective smbd instances.
> 

Yes.

> Each `SmbShare` needs to be associated to one and only one `SambaServer` (or `SambaSecurityDomain`). 

Agree, when talking about a non-clustered scenario. For clustering its a tad bit ambiguous how to describe this as I assume you see the mutliple clustered smbd as sharing one config, right?

> Also very important that (1) each PVC is only attached to only one `SambaServer` (indirectly). If there is a possibility for a subdirectory of a PVC to be turned into another PVC, then it is important that these two PVCs are attached to the _same_ `SambaServer`.

Again, I'm a little unclear on how you think about this in a clustered situation. I envision one Read-Write-Many PVC would be attached to as many pods as needed. I agree that this PVC would be coupled to one (SmbShare, SmbSecurityDoamin) instance, however many smbds/pods that may entail.

gd commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,220 @@
+Spec Options:
+* `mode` - enumerated string "user" or "active-directory" - Indicates that this is an AD joined instance or based on a local mapping of users
+* `users` - mapping - A new subsection pertaining to "user" options - required when mode = "user"
+  * `secret` - string - The name of a secret containing user and group definitions that will be used to create the local users and groups for the service
```
What exactly does the 'secret' string refer to? Is it some sort of embedding user and group definitions in a  cryptographical manner?

phlogistonjohn replied:
In kubernets a secret is a bundle of configuration that may (or may not) be protected by encyption. By default it is simply obscured but you can install add ons that can protect the secret contents more thoroughly.

gd commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,220 @@
+spec:
+  mode: "user"
+  users:
+    secret: "local-users-secret"
```
Just curious: What is that secret used for?

phlogistonjohn replied:
I envision it would contain a json configuration of local users and groups, the actual structure of said json TBD of course. :-)

phlogistonjohn commented:
After a verbal discussion I came to the realization that 'localName" was not appropriate for a CR that would be usable by multiple instances. I have renamed the field "instanceNamePrefix" and updated the description.

phlogistonjohn commented:
As discussed verbally I'd be OK with renaming "SmbSecurityDomain" to "SmbIdentity". It's shorter and is as meaningful as SmbSecurityDomain IMO. Let me know what you think.

obnoxxx commented:
> As discussed verbally I'd be OK with renaming "SmbSecurityDomain" to "SmbIdentity". It's shorter and is as meaningful as SmbSecurityDomain IMO. Let me know what you think.

I think that `SmbIdentity` would really mean that we'd have a 1:1 correspondence between this and an instance. An "Identity" could conceptually by its name not serve as a template for instances.

Maybe call it `SambaSecurityConfig` or `SambaSecuritySetup` instead?

@phlogistonjohn

phlogistonjohn commented:
> > As discussed verbally I'd be OK with renaming "SmbSecurityDomain" to "SmbIdentity". It's shorter and is as meaningful as SmbSecurityDomain IMO. Let me know what you think.
> 
> I think that `SmbIdentity` would really mean that we'd have a 1:1 correspondence between this and an instance. An "Identity" could conceptually by its name not serve as a template for instances.
> 
> Maybe call it `SambaSecurityConfig` or `SambaSecuritySetup` instead?
> 

I'd go with SmbSecurityConfig to match the other CR name. But I'm fine with ...Config as the suffix rather than "Domain".

obnoxxx commented:
> I'd go with `SmbSecurityConfig` to match the other CR name. But I'm fine with ...Config as the suffix rather than "Domain".

Oh, sure, I mistyped Samba for Smb, `SmbSecurityConfig` is good!

phlogistonjohn commented:
I have pushed a new version. 
It changes SmbSecurityDomain to SmbSecurityConfig.
All options pertaining to clustering have been moved under "scaling" and the names are tweaked but  are functionally the same.
Grammar and spelling improvements.
Block formatting on some of the long lines.

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,240 @@
+metadata:
+  name: "cad-files"
+spec:
+  securityDomain: "local-users"
```
`securityConfig`

obnoxxx replied:
```suggestion
  securityConfig: "local-users"
```

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,240 @@
+metadata:
+  name: "documents"
+spec:
+  securityDomain: "cool-domain"
```
`securityConfig`

obnoxxx replied:
```suggestion
  securityConfig: "cool-domain"
```

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,240 @@
+metadata:
+  name: "cad-files"
+spec:
+  securityDomain: "cool-domain"
```
`securityConfig`

obnoxxx replied:
```suggestion
  securityConfig: "cool-domain"
```

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,240 @@
+metadata:
+  name: "documents"
+spec:
+  securityDomain: "local-users"
```
`securityConfig`

obnoxxx replied:
```suggestion
  securityConfig: "local-users"
```

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,240 @@
+metadata:
+  name: "cad-files"
+spec:
+  securityDomain: "local-users"
```
`securityConfig`

obnoxxx replied:
```suggestion
  securityConfig: "local-users"
```

obnoxxx commented on docs/design/crd-proposal-phase1.md:
```diff
@@ -0,0 +1,240 @@
+metadata:
+  name: "documents"
+spec:
+  securityDomain: "cool-domain"
```
`securityConfig`

obnoxxx replied:
```suggestion
  securityConfig: "cool-domain"
```

obnoxxx reviewed (changes requested):
Looks good in general.
A few cases in the examples with the old names.

obnoxxx reviewed (commented):
Oh, and is the duplicate signoff in the commit message intentional?

phlogistonjohn commented:
> Oh, and is the duplicate signoff in the commit message intentional?

Heh. No. It's a side effect of something I am experimenting with.

obnoxxx reviewed (approved):
lgtm
